### ISSUE TITLE
Adaptable rate limit for backfill

### ISSUE THREAD

**Issue by kwannoel** [risingwavelabs/risingwave#16114]
From @chenzl25:
> Assuming that we have a lot of materialized views in the cluster now, and users rely on the freshness of the materialized views to do batch queries. If the user needs to create a new materialized view at this time, our backfilling default strategy will lead to increased barrier latency, and the freshness of the data queried by the user will be reduced. At present, the strategy we give to users is to control the impact of the entire backfill by adjusting the rate limit, but these parameters are a bit too low-level, so I am wondering if we can expose a more high-level policy for users to choose, that is, backfill that affects barrier latency as little as possible. Technically, we can achieve this through adaptive rate limit adjustment (refer to TCP flow control), or by avoiding the impact of the materialized view being created on the existing materialized view through partial checkpoint.

We can do something like slow start (rate_limit=1 for snapshot read side), and increase it gradually, based on barrier collection metrics.

Prerequisite: https://github.com/risingwavelabs/risingwave/issues/16113

**Comment by kwannoel** [comment]
Waiting for rate limit refactor to happen first.

**Comment by kwannoel** [comment]
- [ ] Implement frontend part
- [ ] Implement adaptive rate limit within the backfill executor
- [ ] Add different scenario tests
- [ ] Add metrics for current adapted rate limit

**Comment by maingoh** [comment]
I am wondering if it would be possible to decorrelate the backfill of the new MV from the barrier and the latency of other existing MVs. 
Creating a new MV currently has an impact over other existing streaming jobs (backpressure and increasing latency). I tend to add nodes for the new backfill and I am curious if for the time of the backfill it would not be possible to schedule it on the "empty nodes" in priority so that it does not have an impact on the other real time streaming jobs. When the backfill is finished, the streaming job could be scheduled back in other nodes. There could also be a node pool specific for backfill (just like serving/streaming node pools).
In this case rate limiting would not be as needed and we would not have to find a "good" rate limiting value.

**Comment by chenzl25** [comment]
@maingoh Right, currently we have a feature called snapshot backfill https://arc.net/l/quote/akqsknip to solve this issue. With this feature enable, during the backfilling, the creating MV will be decoupled with the existing streaming jobs. cc @wenym1

**Comment by maingoh** [comment]
Thank you 🙏 I will try it ! Is the default to false because the feature is recent and to avoid breaking existing clusters ?
From a user point of view I would expect it to be enabled by default at some point :)

**Comment by chenzl25** [comment]
It will be changed to true by default in the future, currently it is an experimental feature.

**Comment by maingoh** [comment]
Hello, I have tried to set `streaming_use_snapshot_backfill=true`, but I am getting the following error: `must be either all snapshot_backfill or no snapshot_backfill`. My sources are created via dbt using `table_with_connector` config from kafka. I have other MVs based on those sources.

I am not sure whether it is wanted or a bug, but I feel most users will have sources coming from kafka. Since the evolution of materialized views rely on re-creation from scratch (adding fields) and swapping them, it is important in my opinion to be able to re-create the new materialized view with backfill without impacting the current streaming ones.
